### DISCUSSION TITLE
Expression-constrained parameters change on copy

### DISCUSSION THREAD

**Discussion by TMacioce** [lmfit/lmfit-py#905] [Q&A] [answered]
Hello,

I've run into some unexpected behavior with copying (or pickling/unpickling) `Parameters` objects in the 1.2.2 update. 
In particular, when I perform a fit with an expression constraint and get the output parameters, a copy of those parameters can have slightly different values from the original output. As an example:

```python
import lmfit
import numpy as np

x = np.linspace(0, 1)
y = 2 * x + 1

model = lmfit.models.LinearModel()
model.set_param_hint("total", max=4)
model.set_param_hint("slope", expr="total - intercept")
guess = model.make_params(total=1.5)
fit = model.fit(y, x=x, params=guess)

print("Equal to copy:", fit.params == fit.params.copy())
print("Before:", fit.params.valuesdict())
print("After:", fit.params.copy().valuesdict())
print("total - intercept:", fit.params["total"].value - fit.params["intercept"].value)
# In case it's helpful:
print(fit.fit_report())
```

Output (lmfit=1.2.2):

```
Equal to copy: False
Before: {'slope': 1.9999999552965124, 'intercept': 1.0, 'total': 2.999999999999996}
After: {'slope': 1.999999999999996, 'intercept': 1.0, 'total': 2.999999999999996}
total - intercept: 1.999999999999996
[[Model]]
    Model(linear)
[[Fit Statistics]]
    # fitting method   = leastsq
    # function evals   = 13
    # data points      = 50
    # variables        = 2
    chi-square         = 2.6851e-28
    reduced chi-square = 5.5939e-30
    Akaike info crit   = -3365.83466
    Bayesian info crit = -3362.01061
    R-squared          = 1.00000000
[[Variables]]
    slope:      1.99999996 +/- 1.1357e-15 (0.00%) == 'total - intercept'
    intercept:  1.00000000 +/- 6.5905e-16 (0.00%) (init = 0)
    total:      3.00000000 +/- 6.5905e-16 (0.00%) (init = 1.5)
[[Correlations]] (unreported correlations are < 0.100)
    C(intercept, total) = -0.4848
```

This does not seem to happen on version 1.2.1. Of course, it would be possible to check for approximate equality of the parameters one-by-one using `np.isclose` or something similar. But I think it is convenient to be able to perform exact equality checks on `Parameters`, especially for unit testing, and I would have expected copying to preserve that equality.

Thanks in advance for any thoughts you may have.

**Comment by newville** [comment]
@TMacioce Hm, yeah, that's kind of weird.   I'm not entirely sure what's causing that.  

It looks to me like the first values -- those in `result.params` are slightly off as if the constraint expressions have not been evaluated correctly with the final parameter values.  It may be that these are left over from calculating uncertainties or "ufloats", which is new for 1.2.2.

**Comment by newville** [comment, marked as answer]
@TMacioce  There is a PR (#907) that will fix this.   Basically, the function added in 1.2.2 to create `uncertainty ufloats` for each parameter was not correctly restoring values to the best-fit value, leaving the constrained expressions slightly out-of-sync with the variable parameters.

**Reply by TMacioce** [reply]
Excellent, thanks for looking into this so quickly!
